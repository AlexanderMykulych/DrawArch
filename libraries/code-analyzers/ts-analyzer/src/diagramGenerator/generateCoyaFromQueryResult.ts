import type { QueryResult } from 'neo4j-driver';
import { applyPositioning } from 'coya-core'
import { progressSync } from '../progress/progress';

const getBlockId = (id: string) => `b_${id}`;
const getArrowId = (id: string) => `a_${id}`;

function _generateCoyaFromGraphResult(result: QueryResult) {
  const blocksMap = new Map<string, any>();
  const arrowsMap = new Map<string, any>();
  result
    .records
    .forEach((item) => {
      const graph = item.get('graph');
      graph.relationships.forEach((r) => {
        arrowsMap.set(getArrowId(r.identity), {
          from: getBlockId(r.start.toString()),
          to: getBlockId(r.end.toString()),
          type: 'line',
          label: '',
        });
      });
      graph.nodes.forEach(n => blocksMap.set(getBlockId(n.identity), n));
    });
  // .forEach((item) => {
  //     const from = item.get('from');
  //     const to = item.get('to');
  //     const r = item.get('r');
  //     blocksMap.set(getBlockId(from.identity), from.properties.name);
  //     blocksMap.set(getBlockId(to.identity), to.properties.name);
  //     arrowsMap.set(getArrowId(r.identity), {
  //         from: getBlockId(r.start.toString()),
  //         to: getBlockId(r.end.toString()),
  //         type: 'line',
  //         label: '',
  //     });
  // });

  const blocksEntries = [...blocksMap.entries()];
  const blocks = Object.fromEntries(blocksEntries.map(([key, node]) => [key, node.properties.id]));
  const arrows = Object.fromEntries(arrowsMap.entries());
  const diagram = {
    name: 'coya-ts-analyzer',
    blocks: {
      ...blocks,
      ...arrows,
    },
    phases: [],
    style: {
      blocks: {
        _: {
          css: {
            text: {
              fontSize: 'auto',
            },
          },
        },
        _folder: {
          css: {
            fillStyle: 'solid',
            color: '#ffffffff',
            fill: '#f67c01ff',
          },
        },
        _file: {
          css: {
            fillStyle: 'solid',
            color: '#ffffffff',
            fill: '#008781ff',
          },
        },
        _func: {
          css: {
            fillStyle: 'solid',
            color: '#ffffffff',
            fill: '#2097f3ff',
          },
        },
        ...Object.fromEntries(
          blocksEntries.map(([key, val]) => [key, {
            css: {
              get: val.properties.entityType === 'file'
                ? '_file'
                : val.properties.entityType === 'folder'
                  ? '_folder'
                  : '_func',
            },
          }]),
        ),
      },
    },
  }

  applyPositioning(diagram, 'Dagre')

  return diagram;
}

export const generateCoyaFromGraphResult = progressSync('neo4j data to coya', _generateCoyaFromGraphResult)