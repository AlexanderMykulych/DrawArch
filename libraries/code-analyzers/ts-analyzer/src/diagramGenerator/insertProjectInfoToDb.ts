import { analyzeProject } from "../analysis/project/analyzeProject";
import { useNeo4j } from "../neo4j/useNeo4j";


export async function insertProjectInfoToDb(path: string) {
  const codeInfos = await analyzeProject(path);

  const db = useNeo4j();
  await db.init();
  await db.insertCodeInfos(codeInfos);

  return {
    db,
    codeInfos,
  }
}
