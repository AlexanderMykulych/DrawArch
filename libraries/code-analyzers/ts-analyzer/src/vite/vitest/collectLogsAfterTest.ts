import { ActionEntity, ActionEntityType, CodeInfo, CodeInfoType, Relationship, RelationType } from "../../analysis/types"
import { useNeo4j } from "../../neo4j/useNeo4j"
import type { LogItem, StartLogItem } from "../shim"


export async function collectLogsAfterTest(insertToDb: true): Promise<void>
export async function collectLogsAfterTest(): Promise<CodeInfo[]>
export async function collectLogsAfterTest(insertToDb?: true): Promise<CodeInfo[] | void> {
  if (!globalThis.__getLogs) {
    return
  }

  const logs = globalThis.__getLogs()

  const codeInfos = logsToCodeInfo(logs)

  if (insertToDb) {
    const db = useNeo4j()
    await db.insertCodeInfos(codeInfos);
  } else {
    return codeInfos
  }
}


export function logsToCodeInfo(logs: LogItem[]): CodeInfo[] {
  return logs
    .flatMap<CodeInfo>((item, index, items) => {
      const result = []
      if (item.type === 'end') {
        const callEntity: ActionEntity = {
          type: CodeInfoType.Entity,
          entityType: ActionEntityType.Call,
          id: getCallActionId(item, index),
          executionTime: item.execTime,
        }
        const callRelation: Relationship = {
          type: CodeInfoType.Relationship,
          relationType: RelationType.CallEntity,
          from: callEntity.id,
          to: item.id,
        }

        result.push(callEntity, callRelation)

        const [parentStartItem, parentIndex] = findParentStartLogFromIndex(items, index, item.id)

        if (parentStartItem) {
          const parentRelation: Relationship = {
            type: CodeInfoType.Relationship,
            relationType: RelationType.Parent,
            from: getCallActionId(parentStartItem, parentIndex),
            to: callEntity.id,
          }
          result.push(parentRelation)
        }
      }
      return result
    })
}

export function findStartLogFromIndex(items: LogItem[], index: number, id: string): StartLogItem {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id === id) {
      return startItem
    }
  }
  throw 'Can not find start element for: ' + id
}

export function findParentStartLogFromIndex(items: LogItem[], index: number, id: string): [StartLogItem, number] | [null, null] {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id !== id) {
      return [startItem, i]
    }
  }

  return [null, null]
}

export function getCallActionId(item: LogItem, index: number): string {
  // return `call:${item.id}`
  return `call[${index}]:${item.id}`
}