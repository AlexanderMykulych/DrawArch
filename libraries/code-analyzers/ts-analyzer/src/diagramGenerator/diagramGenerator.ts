import { insertProjectInfoToDb } from "./insertProjectInfoToDb";

export async function diagramGenerator(path: string) {
  await insertProjectInfoToDb(path)

}
