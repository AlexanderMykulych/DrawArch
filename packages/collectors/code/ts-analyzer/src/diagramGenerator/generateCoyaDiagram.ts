import { generateCoyaFromGraphResult } from "./generateCoyaFromQueryResult";
import { insertProjectInfoToDb } from "./insertProjectInfoToDb";

export async function generateCoyaDiagram(path: string) {
  const { db } = await insertProjectInfoToDb(path)

  const graph = await db.read(`
  match p = (n1)-[*..{relationType: "parent"}]->(n2)
  where
    not n1.id contains 'node_modules/'
    and
    not n2.id contains 'node_modules/'
  return apoc.agg.graph(p) as graph
`)
  // const graph = await db.read('match p = (n1)-[*0..]->() return apoc.agg.graph(p) as graph')

  if (!graph) {
    throw 'graph is empty'
  }

  const coya = generateCoyaFromGraphResult(graph)

  return {
    db,
    coya,
  }
}
