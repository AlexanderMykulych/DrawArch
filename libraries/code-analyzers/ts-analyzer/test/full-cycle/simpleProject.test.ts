import path from 'path'
import { test, expect, describe } from 'vitest'
import { insertProjectInfoToDb } from '../../src/diagramGenerator/insertProjectInfoToDb'

describe.each([
  {
    projectPath: '/cases/01_simple',
  },
  {
    projectPath: '/cases/02_function',
  },
  {
    projectPath: '/cases/03_function_relation',
  },
  // {
  //   projectPath: '/cases/04_vue',
  // },
])('project: $projectPath', async ({ projectPath }) => {
  const fullProjectPath = path.join(__dirname, projectPath)

  const { db } = await insertProjectInfoToDb(fullProjectPath)

  test(`should return project info entities from db: ${projectPath}`, async () => {

    const result = await db.read('match(e) return e.id')

    expect(result?.records.length).greaterThan(0)

  })

  test(`should return project info graph from db: ${projectPath}`, async () => {

    const result = await db.read('match(n1)-[r]->(n2) return n1.id, r.relationType, n2.id')

    expect(result?.records.length).greaterThan(0)

  })
})
