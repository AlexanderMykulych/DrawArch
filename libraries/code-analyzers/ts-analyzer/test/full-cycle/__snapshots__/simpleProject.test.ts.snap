// Vitest Snapshot v1

exports[`project: /cases/01_simple > should return project info entities from db: /cases/01_simple 1`] = `
[
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/add.ts/add",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/add.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
]
`;

exports[`project: /cases/01_simple > should return project info graph from db: /cases/01_simple 1`] = `
[
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/add.ts",
      "parent",
      "/add.ts/add",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "import",
      "/add.ts",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
]
`;

exports[`project: /cases/02_function > should return project info entities from db: /cases/02_function 1`] = `
[
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts/func1",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts/func2",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
]
`;

exports[`project: /cases/02_function > should return project info graph from db: /cases/02_function 1`] = `
[
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "parent",
      "/index.ts/func2",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "parent",
      "/index.ts/func1",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
]
`;

exports[`project: /cases/03_function_relation > should return project info entities from db: /cases/03_function_relation 1`] = `
[
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts/func1",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts/func3",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts/func4",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/funcs.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts/func2",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/funcs.ts/func2",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/funcs.ts/func3",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
]
`;

exports[`project: /cases/03_function_relation > should return project info graph from db: /cases/03_function_relation 1`] = `
[
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts/func1",
      "use",
      "/index.ts/func2",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts/func3",
      "use",
      "/index.ts/func2",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts/func4",
      "use",
      "/index.ts/func3",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts/func4",
      "use",
      "/index.ts/func2",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/funcs.ts",
      "parent",
      "/index.ts/func3",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/funcs.ts",
      "parent",
      "/index.ts/func2",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/funcs.ts",
      "parent",
      "/funcs.ts/func3",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/funcs.ts",
      "parent",
      "/funcs.ts/func2",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "parent",
      "/index.ts/func4",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "parent",
      "/index.ts/func3",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "parent",
      "/index.ts/func1",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "import",
      "/funcs.ts",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
]
`;

exports[`project: /cases/04_vue > should return project info entities from db: /cases/04_vue 1`] = `
[
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/cmp.vue.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/utils.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/node_modules/typescript/lib/lib.es5.d.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "vue",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/cmp.vue.ts/render",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/cmp.vue.ts/setup",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/utils.ts/utilFn",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/index.ts/cmpFunc",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/cmp.vue.ts/openBlock",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/cmp.vue.ts/createElementBlock",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/cmp.vue.ts/toDisplayString",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/cmp.vue.ts/ref",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
  Record {
    "_fieldLookup": {
      "e.id": 0,
    },
    "_fields": [
      "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    ],
    "keys": [
      "e.id",
    ],
    "length": 1,
  },
]
`;

exports[`project: /cases/04_vue > should return project info graph from db: /cases/04_vue 1`] = `
[
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts",
      "parent",
      "/cmp.vue.ts/setup",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts",
      "parent",
      "/cmp.vue.ts/render",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts",
      "import",
      "/utils.ts",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts",
      "import",
      "vue",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/utils.ts",
      "parent",
      "/utils.ts/utilFn",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/node_modules/typescript/lib/lib.es5.d.ts",
      "parent",
      "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/node_modules/typescript/lib/lib.es5.d.ts",
      "parent",
      "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "parent",
      "/index.ts/cmpFunc",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/index.ts",
      "import",
      "/cmp.vue.ts",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts/render",
      "use",
      "/cmp.vue.ts/createElementBlock",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts/render",
      "use",
      "/cmp.vue.ts/toDisplayString",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts/render",
      "use",
      "/cmp.vue.ts/openBlock",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts/setup",
      "use",
      "/utils.ts/utilFn",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts/setup",
      "use",
      "/node_modules/typescript/lib/lib.es5.d.ts/defineProperty",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts/setup",
      "use",
      "/node_modules/typescript/lib/lib.es5.d.ts/Object",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
  Record {
    "_fieldLookup": {
      "n1.id": 0,
      "n2.id": 2,
      "r.relationType": 1,
    },
    "_fields": [
      "/cmp.vue.ts/setup",
      "use",
      "/cmp.vue.ts/ref",
    ],
    "keys": [
      "n1.id",
      "r.relationType",
      "n2.id",
    ],
    "length": 3,
  },
]
`;
